### ISSUE TITLE
Dump model into upstream and downstream

### ISSUE THREAD

**Issue by ghost** [idiap/apam#2]
Hi there
Because I would like to make our model could support online decoding with Kaldi.
I am curious that is it possible to separate the model (trained by PyTorch) into upstream model and downstream model (like `final.mdl` in kaldi format)?

**Comment by mrsrikanth** [comment]
Sorry. While it is possible to save the upstream and downstream models separately, it is not possible to convert them to Kaldi format.
Converting the downstream model would require writing a TDNN-F class in Kaldi. You cannot reuse the existing TDNN-F implementation in Kaldi because ours is slightly different in the way contexts are used.

**Comment by ghost** [comment]
Hi 
Is that possible to use torch script to convert the upstream/downstream model to some specific format that it could be loaded in C++

I tried `up = torch.jit.trace(self.upstream, feats)` but it failed.

**Comment by mrsrikanth** [comment]
I am not familiar with how to use torch.jit, but perhaps you can try converting it with onnx. Note that even if you converted the TDNN-F layers in the downstream model, it would not be compatible with Kaldi due to how we handle the context.

If your requirement is an online decoder, I could wrap the decoder in pkwrap (a dependency for this package).

**Comment by ghost** [comment]
Yes. My goal is to apply this AM structure to original online decoder.
Sounds great that we could use online decoder in pkwrap! Is that means it could act like online decoding like Kaldi but using
self-supervised model?

BTW,
The reason why I would like to study torch script is that it might work if I replace the the `am_nnet` of `nnet3::DecodableNnetSimpleLoopedInfo` with torch jit model?

thank you!

**Comment by mrsrikanth** [comment]
pkwrap doesn't support it yet. I can add it if it is a requirement. Although I need some more information on which online decoder you are referring to and how you would normally use it with a Kaldi model.

Note that forward pass through pytorch will be done at one-shot (hence, won't necessarily be online).

**Comment by ghost** [comment]
Usually I use `online2-wav-nnet3-latgen-faster` to decode our test data.
I thought the feature streaming part is completed at `OnlineNnet2FeaturePipeline` and we only take the chunk as `am_nnet` input in `SingleUtteranceNnet3Decoder`.
That means if we could use torch script to call it from C++, then we only send the chunk to torch model and receive the acoustic score. So we might avoid the layer compatible issue?

**Comment by mrsrikanth** [comment]
The idea is correct: override Kaldi's computation of network output for each chunk. To be honest, I haven't worked with torch script, so I couldn't tell you how to use that. But it seems doable through the existing C++ interface in pkwrap.

**Comment by ghost** [comment]
Thanks!
But now I get many problems from exporting the model to torch script.
It looks like the `fast-transformer` is not scriptable and needs to rewrite its source code about mask part.

If my requirement is integrating our `apam` model in Kaldi online decoding, what else methods that you recommend?

BTW,
Does pkwrap team plan to export model to torch script or ONNX in future?

**Comment by mrsrikanth** [comment]
> If my requirement is integrating our apam model in Kaldi online decoding, what else methods that you recommend?

One way would be to wrap the online decoder in pkwrap so that they can be called directly from python. Other method can be quite tedious to develop and use.

> Does pkwrap team plan to export model to torch script or ONNX in future?

At the moment it may not be easy to export them to ONNX because of the many dependencies. I'll have to look into what it takes to make the models exportable.
